third_party/move/move-model/bytecode/abstract_domain_derive/src/lib.rs

```diff
@@ -0,0 +1,114 @@
+//! Derive macro for `AbstractDomain`
+//!
+//! Currently we can only derive for structs.
+//! For tuple structs, the derived join joins each field;
+//! for structs with named fields, the derived join joins each field with #[join] attribute.
+
+use proc_macro::TokenStream;
+use quote::{quote, ToTokens};
+use syn::{self, parse_macro_input, DeriveInput, Fields};
+
+/// Given a field name, generates TokenStream of
+/// `join_result = JoinResult::combine(join_result, self.field_name.join(&other.field_name));`
+fn gen_join_field(field: proc_macro2::TokenStream) -> proc_macro2::TokenStream {
+    quote! {
+        join_result = JoinResult::combine(join_result, self.#field.join(&other.#field));
+    }
+}
+
+#[proc_macro_derive(AbstractDomain, attributes(join))]
+/// Derives `AbstractDomain` for structs. The derived `join` method joins selected fields of a struct, or all fields for structs with anonymous fields, and returns the combined join results.
+/// The joined fields must implement `AbstractDomain`.
+/// # Usage
```

Comment:
Consider adding the precondition that each constituent field type itself have implemented `AbstractDomain`.

Reply:
This is already documented?

Reply:
Must have missed it when I read it.